### PR TITLE
add model num params display, gpu memory metrics

### PR DESCRIPTION
**Pull request by lessw2020**
This PR is the start of adding perf related metrics. 
1 - This PR adds function for logging the total num of unique model params, with option for only counting trainable params as well. (for future peft/qlora type work).
2 - logs it with comma formatted logging and model name ala:
Screenshot 2024-02-12 at 4 12 22 PM

this helps de-mistify for example the size of our debug model as well:
Screenshot 2024-02-12 at 4 10 17 PM

**additional updates** - added in gpu mem tracking.  We want to show the user peak memory stats, as well as monitor and alert for any cudacachealloc retries which are a perf hindrance. 

Thus, added class GPUMemoryMonitor:
usage:
1 - instantiate
Screenshot 2024-02-13 at 9 32 11 AM

2 - start of training = start_monitoring()
3 - end of training = stop_monitoring()
4 - show results = get_peak_stats_str() and rank0_log it.
Screenshot 2024-02-13 at 9 12 45 AM

### CODE REVIEW

File: torchtrain/train_configs/train_config.toml [hunk 1]
```diff
@@ -3,7 +3,7 @@
 dump_folder = "./torchtrain/outputs"
 
 [profiling]
-run_profiler = true
+run_profiler = false
```
Comment (wanchaol) [inline torchtrain/train_configs/train_config.toml]: why turning the profiling off?

File: torchtrain/metrics_utils.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+    def get_current_stats(self):
+        self.device_memory_usage = torch.cuda.memory_allocated(self.device)
+        self.device_memory_usage_gb = format_to_gb(self.device_memory_usage)
+        self.device_memory_utilization = convert_to_gpu_pct(
+            self.device_memory_usage, self.device_capacity
+        )
+        self.device_memory_utilization_str = f"{self.device_memory_utilization:.2f}%"
+
+    def start_monitoring(self):
```
Comment (wanchaol) [inline torchtrain/metrics_utils.py]: do we need this start monitoring/stop monitoring method?

IMO the constructor reset the memory stats already, the only method we would need in the train loop is sth like get_peak_memory_stats?

File: run_llama_train.sh [hunk 3]
```diff
@@ -22,8 +23,10 @@ CHECKPOINT_FOLDER=${CHECKPOINT_FOLDER:-""}
...
-torchrun --nproc_per_node=${NGPU} \
+torchrun --nproc_per_node=${NGPU} --rdzv_endpoint="localhost:5972" \
 --local-ranks-filter ${LOG_RANK} --role rank --tee 3 \
-train.py --steps 10 --compile \
---pp_degree ${PP} --sp_degree ${SP} --dp_degree ${DP}
---checkpoint-folder=${CHECKPOINT_FOLDER} --checkpoint-interval=${CHECKPOINT_INTERVAL}
+train.py --steps 10 \
+--model ${MODEL} --model_conf ${MODEL_CONF} \
+--pp_degree ${PP} --sp_degree ${SP} --dp_degree ${DP} \
+# --compile \
```
Comment (wanchaol) [inline run_llama_train.sh]: similar comment to the other PRs on this shell script changes

File: torchtrain/metrics_utils.py (hunk 2 above)
Comment (lessw2020) [reply torchtrain/metrics_utils.py]: yes, that's a good simplification.  Will update to do that.

**Pull request merged**